core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java

```diff
@@ -8112,6 +8112,11 @@ public void _testGroupExpressionEquivalenceParams() {
         .type("RecordType(VARCHAR(10) OA) NOT NULL");
   }
 
+  @Test public void testItemOperatorException() {
+    sql("select ^name^[0] from dept")
+        .fails("Cannot apply item operator to 'VARCHAR' for field 'DEPT.NAME'");
```

Comment:
I'm afraid the message is not quite understandable.
It is way better than the original, however it is still confusing.

How the error message would like for case like  `select ^(2+2)^[0] from dept`?

Reply:
yes,in the case like `select ^(2+2)^[0] from dept`, the error message is confusing.

Reply:
I think error message like` Cannot apply item operator to 'VARCHAR' type` is enough.